Utilize parser-grammar for the remaining background related properties

weinig opened the pull request:
#### d2163c292b7842958e02cda6bd160ccbeab80e68
<pre>
Utilize parser-grammar for the remaining background related properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=249493">https://bugs.webkit.org/show_bug.cgi?id=249493</a>
rdar://103455420

Reviewed by Tim Nguyen.

The remaining background / mask properties can be parsed using just
three custom parts: &lt;position-x&gt;, &lt;position-y&gt; and &lt;repeat-style&gt;.
Utilize that to adopt parser-grammar for these remaining properties.

* Source/WebCore/css/CSSBackgroundRepeatValue.cpp:
(WebCore::CSSBackgroundRepeatValue::CSSBackgroundRepeatValue):
* Source/WebCore/css/CSSBackgroundRepeatValue.h:
Add ability to create from CSSValueIDs.

* Source/WebCore/css/CSSProperties.json:
Update to use parser-grammar in more places.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeBackgroundComponent):
Moved from CSSPropertyParserHelpers.cpp.

(WebCore::CSSPropertyParser::parseShorthand):
This was the last caller of consumeCommaSeparatedBackgroundComponent
so we can just inline the implementation.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumePositionX):
(WebCore::CSSPropertyParserHelpers::consumePositionY):
(WebCore::CSSPropertyParserHelpers::consumeRepeatStyle):
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitBackgroundSize):
Update signature to allow being called from generated parsers.

(WebCore::CSSPropertyParserHelpers::consumeRepeatStyleComponent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundRepeat): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundPositionX): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundPositionY): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleMaskRepeat): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitMaskPositionX): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitMaskPositionY): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundComponent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeCommaSeparatedBackgroundComponent): Deleted.
Remove no longer needed helpers.

Canonical link: <a href="https://commits.webkit.org/258046@main">https://commits.webkit.org/258046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/228e90f707b9352e2f69c98ace4655fc0b88f11f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100662 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9807 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33703 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109964 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170236 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104647 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10742 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/469 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93063 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107809 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106443 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8109 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91368 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34728 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90026 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22770 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77709 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3507 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24286 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3526 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9638 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43783 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5524 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5313 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -9237,6 +9217,26 @@
                 "url": "https://www.w3.org/TR/css-backgrounds-3/#background-image"
             }
         },
+        "<single-background-repeat>": {
+            "grammar": "<repeat-style>",
+            "exported": true,
+            "specification": {
+                "category": "css-backgrounds",
+                "url": "https://www.w3.org/TR/css-backgrounds-3/#background-repeat"
+            }
+        },
+        "<single-background-position-x>": {
+            "grammar": "<position-x>",
+            "exported": true,
+            "status": "non-standard",
+            "comment": "Internally used to represent horizontal part of standard <bg-position>"
+        },
+        "<single-background-position-y>": {
+            "grammar": "<position-y>",
+            "exported": true,
+            "status": "non-standard",
+            "comment": "Internally used to represent vertical part of standard <bg-position>"
+        },
```
Can we define `<repeat-style>`/`<position-x>`/`<position-y>` from `CSSProperties.json` as well?

weinig replied:
Not at the moment. No.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -6070,7 +6054,7 @@
                     "background-size"
                 ],
-                "custom-parser": true
+                "parser-grammar": "<single-webkit-background-size>#"
```
question: Does this generate any new code that might interfere with the legacy shorthand behavior? (e.g. calling `addProperty()` with `CSSPropertyBackgroundSize`)

nt1m replied:
(This should be caught by test failures, but asking just in case)

weinig replied:
It shouldn't. Shorthands are skipped currently.

Commit:
Utilize parser-grammar for the remaining background related properties
https://bugs.webkit.org/show_bug.cgi?id=249493
rdar://103455420

Reviewed by Tim Nguyen.

The remaining background / mask properties can be parsed using just
three custom parts: <position-x>, <position-y> and <repeat-style>.
Utilize that to adopt parser-grammar for these remaining properties.

* Source/WebCore/css/CSSBackgroundRepeatValue.cpp:
(WebCore::CSSBackgroundRepeatValue::CSSBackgroundRepeatValue):
* Source/WebCore/css/CSSBackgroundRepeatValue.h:
Add ability to create from CSSValueIDs.

* Source/WebCore/css/CSSProperties.json:
Update to use parser-grammar in more places.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeBackgroundComponent):
Moved from CSSPropertyParserHelpers.cpp.

(WebCore::CSSPropertyParser::parseShorthand):
This was the last caller of consumeCommaSeparatedBackgroundComponent
so we can just inline the implementation.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumePositionX):
(WebCore::CSSPropertyParserHelpers::consumePositionY):
(WebCore::CSSPropertyParserHelpers::consumeRepeatStyle):
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitBackgroundSize):
Update signature to allow being called from generated parsers.

(WebCore::CSSPropertyParserHelpers::consumeRepeatStyleComponent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundRepeat): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundPositionX): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundPositionY): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleMaskRepeat): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitMaskPositionX): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitMaskPositionY): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundComponent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeCommaSeparatedBackgroundComponent): Deleted.
Remove no longer needed helpers.

Canonical link: https://commits.webkit.org/258046@main
